src/spacepackets/ccsds/tm_frame.py

```diff
@@ -92,48 +126,81 @@ def unpack(cls, data: bytes) -> TransferFrameSecondaryHeader:
 class TmTransferFrame:
     def __init__(
         self,
+        length: int,
         primary_header: TmFramePrimaryHeader,
         secondary_header: Optional[TransferFrameSecondaryHeader],
         data_field: bytes,
         op_ctrl_field: Optional[bytes],
         frame_error_control: Optional[bytes],
     ) -> None:
+        if length < 2048:  # According to CCSDS TM SYNCHRONIZATION AND CHANNEL CODING Blue Book
+            self.length = length
+        else:
+            raise ValueError(f"Tm frame length {length} exceeds maximum of 2048.")
         self.primary_header = primary_header
         self.secondary_header = secondary_header
         self.data_field = data_field
         self.op_ctrl_field = op_ctrl_field
         self.frame_error_control = frame_error_control
 
+    def pack(self, data: TmTransferFrame) -> bytes:
+        packed = bytearray()
+        packed.extend(data.primary_header.pack(data.primary_header))
+        if data.primary_header.frame_datafield_status.secondary_header_flag:
+            packed.extend(data.secondary_header.pack(data.secondary_header))
+        packed.extend(data.data_field)
+        if data.op_ctrl_field is not None:
+            packed.extend(data.op_ctrl_field)
+        if data.frame_error_control is not None:
+            packed.extend(data.frame_error_control)
+        if len(packed) != data.length:
+            raise ValueError(
+                f"Transfer frame length {len(packed)} not equal to expected length {data.length}"
+            )
+        return bytes(packed)
+
     @classmethod
-    def unpack(cls, raw_frame: bytes, has_error_control_field: bool) -> TmTransferFrame:
+    def unpack(
+        cls, raw_frame: bytes, length: int, has_error_control_field: bool
```

Comment:
if the length, and the raw datastream the frame is deserialized from are two different things, I would re-name `raw_frame` to `data`. I guess it makes sense to pass length separately because this frame does not have a length field.